code/drasil-code/lib/Language/Drasil/Choices.hs

```diff
@@ -362,4 +366,27 @@ choicesSent chs = map chsFieldSent [
 
 -- | Helper to combine pairs of 'Sentence's for rendering 'Choices'.
 chsFieldSent :: (Sentence, Sentence) -> Sentence
-chsFieldSent (rec, chc) = rec +:+ S "selected as" +:+. chc
\ No newline at end of file
+chsFieldSent (rec, chc) = rec +:+ S "selected as" +:+. chc
+
+-- | List of user defined function names.
+-- | List is populated with default values.
+fnList :: [(InternalConcept, Name)]
+fnList = [
+  (GetInput, "get_input"), 
+  (DerivedValues, "derived_values"), 
+  (InputConstraints, "input_constraints"), 
+  (WriteOutput, "write_output")]
+
+-- | Data type of user defined concepts
+data InternalConcept = InputConstraints | WriteOutput | DerivedValues 
+  | GetInput | InputParameters | InputFormat deriving Eq  
+
+-- | Returns user defined function Name
+genICFuncName :: InternalConcept -> Name
+genICFuncName ic = existsFuncName(lookup ic (functionNames defaultChoices))
```

Comment:
I _think_ the `defaultChoices` should be replaced with a function argument in `genICFuncName`, or else the Choices option you added would always be overridden by the basic `defaultChoices` Choices.

Reply:
@balacij I'm not sure I follow. `functionNames = fnList` would be overridden by a basic `defaultChoices` Choices? Why would this happen? Could you also please elaborate on the proposed solution to fix such an issue?

Reply:
By "basic `defaultChoices`", I meant the `Choices` value that we have defined and commonly use -- `defaultChoices` is the variable containing it.

You added a mapping of InternalConcepts to Names in the `Choices` record, but, when grabbing the names, I _think_ this is pulling from the built-in `defaultChoices` instead of the one the user provided, which is why I suggested that it might need to be an argument, so that it can grab from the `Choices` instance the user provided instead of strictly. Otherwise, the user provided information for the function names wouldn't be used.

Reply:
> Are the built-in defaultChoices not overwritten by the user's Choices?

`defaultChoices` (and all data in Haskell) is immutable. The "dynamic"/mutated data has to be provided through the input arguments, unless, I believe, you're using `ST` (where we can have mutable variables, and which I don't think we want to use).

Sounds good to me! :smile: